Implement bytecode callbacks with static, never-modified bytecode

xavierleroy opened the pull request:
This is an alternative to #13549.  See this PR for the detailed description of the problem that needs to be fixed.

Like #13549, the present PR avoids modifying bytecode dynamically, and making it thread-local.

Unlike #13549, it preserves the ability to perform a callback on N arguments in one go, instead of having to cut it in slices of 1 to 3 arguments.

xavierleroy committed:
interp.c: more general API for the bytecode API

Support passing the initial values of env and extra_args as parameters
to the bytecode interpretation function.

xavierleroy committed:
Reimplement bytecode callbacks without per-thread bytecode fragments

- Use the new `caml_bytecode_interpreter` API to jump straight to the
  function's code.
- Avoid modifying bytecode in place.
- Avoid per-thread bytecode.
- Register bytecode early and only once.

Fixes:  #13402
Fixes:  #13512
Closes: #13549

dustanddreams reviewed (approved):
This also makes the failures disappear and looks good to me.

stedolan commented:
This looks good! This approach does seem cleaner than either modifying bytecode or slicing up callbacks into 3-argument chunks, so I think this is a better fix than #13549.

However, I think there's a bug in the current version: the old code, going through `Instruct(APPLY)`, used to do a `goto check_stacks` before executing the code from the callback. I think this stack check is important, as I think it's possible to reach caml_callback without much available stack (there's no stack check before `Instruct(C_CALLn)`, so you can enter C code having already consumed part of your stack).

I think there's a simple fix: the interpreter should do a `goto check_stacks` right before entering the main loop, so that a stack check occurs before any instructions are executed.

shindere commented:
Stephen Dolan (2024/10/16 02:36 -0700):
> I think there's a simple fix: the interpreter should do a `goto
> check_stacks` right before entering the main loop, so that a stack
> check occurs before any instructions are executed.

Won't that be too often then? I suppose many instructions do actually
not change the stack and it then feels a pity to me to do a check before
they are executed, leading to a performance loss. Or am I missing
something maybe?

stedolan commented:
I'm not suggesting a check inside the loop, before every instruction - that would indeed cause performance loss. I'm suggesting a single check, _before_ the loop is entered, which would be once per invocation of `caml_bytecode_interpreter`.

shindere commented:
Stephen Dolan (2024/10/16 05:55 -0700):
> I'm not suggesting a check inside the loop, before every instruction -
> that would indeed cause performance loss. I'm suggesting a single
> check, _before_ the loop is entered, which would be once per
> invocation of `caml_bytecode_interpreter`.

Ah okay, sorry I misunderstood. And you think such a check would be
enough, then?

NickBarnes commented:
> I think this stack check is important, as I think it's possible to reach caml_callback without much available stack (there's no stack check before `Instruct(C_CALLn)`, so you can enter C code having already consumed part of your stack).

On this subject, shouldn't `caml_callbackN_exn` do a stack limit check before subtracting `narg + 4` from `sp`? Maybe I'm missing something obvious about the bytecode stacks.

xavierleroy commented:
> the interpreter should do a goto check_stacks right before entering the main loop

Well spotted!  See the latest commit on this PR.

> shouldn't caml_callbackN_exn do a stack limit check before subtracting narg + 4 from sp

Right now we  rely on the "red zone" at the bottom of the stack to absorb the arguments.  There's an assertion but it's not up to date (it assumes 256 words of red zone while it's only 32 words nowadays).  Maybe it would be better to check and resize the stack in caml_callbackN_exn.  Will look into it in a couple of days.

NickBarnes commented:
Ah, I see, `Stack_threshold_words` ?

xavierleroy committed:
Bytecode callbacks: make sure there is enough stack space

... before copying the arguments to the stack and calling the bytecode
interpreter.

xavierleroy commented:
Yes, `Stack_threshold_words`.

I pushed an alternative to @stedolan's suggestion, based on @NickBarnes' remark, where `caml_callbackN` checks for stack space and reallocates the stack if needed.  I doubt the test suite exercises this code, though.  I'll try to come up with a test later this week.

xavierleroy committed:
Test callbacks with many arguments

For bytecode, this tests stack resizing in caml_callbackN.

xavierleroy committed:
Changes entry for #13553

xavierleroy commented:
I added a test that triggers the stack resizing, and a Changes entry.  This PR is ready for a final review.

dustanddreams commented on runtime/callback.c:
```diff
@@ -83,54 +83,54 @@ Caml_inline void restore_stack_parent(caml_domain_state* domain_state,
-  domain_state->current_stack->sp -= narg + 4;
+  /* Ensure there's enough stack space */
+  intnat req = narg + 3 + Stack_threshold_words;
+  if (domain_state->current_stack->sp - req <
```
This almost calls for an internal version of `caml_ensure_stack_capacity`, but that can be done later if worth doing.

jmid commented:
After a good round of testing this afternoon I can confirm that this fixes both #13512 and #13402.
Thank you both!

xavierleroy commented:
Cleaned up the history.  Rebased to fix conflict.  Knelt in front of check-typo.  Time to merge!

Octachron commented:
Cherry-picked to the 5.2 branch as 71d85e068d in prevision of the upcoming 5.2.1 release.
